beacon/validator/block_builder.go

```diff
@@ -214,15 +214,11 @@ func (s *Service) retrieveExecutionPayload(
 	parentBlockRoot common.Root,
 	slotData *types.SlotData,
 ) (ctypes.BuiltExecutionPayloadEnv, error) {
-	//
 	// TODO: Add external block builders to this flow.
 	//
 	// Get the payload for the block.
-	envelope, err := s.localPayloadBuilder.RetrievePayload(
-		ctx,
-		slotData.GetSlot(),
-		parentBlockRoot,
-	)
+	slot := slotData.GetSlot()
+	envelope, err := s.localPayloadBuilder.RetrievePayload(ctx, slot, parentBlockRoot)
 	if err == nil {
```

Comment:
@calbera question: shouldn't we carry out `ProcessFork` as well, using `envelope.GetExecutionPayload().GetTimestamp()`?
`ProcessFork` does modify the state, but with the code as is, we are doing it some of the times but not others.

Reply:
Not necessary in this case because 
- here the optimistically built envelope is returned
- back into control flow in BuildBlockAndSidecars
- then comuteAndSetStateRoot is called which then invokes state processor Transition function, which then calls in the following order:
  - ProcessSlots (no-op in this case since ProcessSlots was called before even retrieving the envelope)
  - ProcessFork
  - ProcessBlock

Reply:
But then why is `ProcessFork` needed at all here? It does not seem to me that `s.localPayloadBuilder.RequestPayloadSync` need the state to be prepared with `ProcessFork`

Reply:
In the case there is no optimistically built payload, we will not return the envelope here. Continuing in this function, before we send an FCU to EL to build this new payload we should process the fork in case this new payload is entering a new fork version. Before the FCU call, there are a couple reads from the beacon state which may be fork version dependent.

Reply:
You're right